ui/src/containers/VolumePage.js

```diff
@@ -0,0 +1,750 @@
+  useRefreshEffect(refreshVolumesAction, stopRefreshVolumesAction);
+  useRefreshEffect(
+    refreshPersistentVolumesAction,
+    stopRefreshPersistentVolumesAction,
+  );
+  useEffect(() => {
+    dispatch(fetchPodsAction());
+    dispatch(refreshVolumeStatsAction());
+    dispatch(fetchPersistentVolumeClaimAction());
+  }, [dispatch]);
+  useEffect(() => {
+    dispatch(refreshAlertsAction());
+    return () => dispatch(stopRefreshAlertsAction());
+  }, [dispatch]);
+
+  const theme = useSelector((state) => state.config.theme);
+  const pods = useSelector((state) => makeGetPodsFromUrl(state, props));
+  const node = useSelector((state) => makeGetNodeFromUrl(state, props));
+  const volumes = useSelector((state) => makeGetVolumesFromUrl(state, props));
+  const pVList = useSelector((state) => state.app.volumes.pVList);
+  const pVCList = useSelector((state) => state.app.volumes.pVCList);
+  const alerts = useSelector((state) => state.app.monitoring.alert);
+
+  const volumeUsedList = useSelector(
+    (state) => state.app.monitoring.volumeStats.volumeUsed,
+  );
+  const volumeUsedCurrentList = useSelector(
+    (state) => state.app.monitoring.volumeStats.volumeUsedCurrent,
+  );
+  const volumeThroughputWriteList = useSelector(
+    (state) => state.app.monitoring.volumeStats.volumeThroughputWrite,
+  );
+  const volumeThroughputReadList = useSelector(
+    (state) => state.app.monitoring.volumeStats.volumeThroughputRead,
+  );
+  const volumeLatencyList = useSelector(
+    (state) => state.app.monitoring.volumeStats.volumeLatency,
+  );
+  const volumeLatencyCurrent = useSelector(
+    (state) => state.app.monitoring.volumeStats.volumeLatencyCurrent,
+  );
+  const volumeIOPSReadList = useSelector(
+    (state) => state.app.monitoring.volumeStats.volumeIOPSRead,
+  );
+  const volumeIOPSWriteList = useSelector(
+    (state) => state.app.monitoring.volumeStats.volumeIOPSWrite,
+  );
+  const queryStartingTime = useSelector(
+    (state) => state.app.monitoring.volumeStats.queryStartingTime,
+  );
+
+  // Should be the selected volume from the volume list
+  const [currentVolumeName, setCurrentVolumeName] = useState(
+    match.params.volumeName,
+  );
+  const [searchedVolumeName, setSearchedVolumeName] = useState('');
+
+  const pV = pVList.find((pv) => pv.metadata.name === currentVolumeName);
+  const volume = volumes.find(
+    (volume) => volume.metadata.name === currentVolumeName,
+  );
+
+  const volumeStatus = computeVolumeGlobalStatus(
+    volume?.metadata?.name,
+    volume?.status,
+  );
+
+  // in order to get the used pod(s)
+  const PVCName = pV?.spec?.claimRef?.name;
+  const UsedPod = pods.find((pod) =>
+    pod.volumes.find((volume) => volume.persistentVolumeClaim === PVCName),
+  );
+
+  // get the alert base on the current
+  const alertlist = alerts?.list?.filter(
+    (alert) => alert.labels.persistentvolumeclaim === PVCName,
+  );
+
+  // Data for volume table
+  const volumeListData = volumes.map((volume) => {
```

Comment:
This one is a bit tricky, we might want to use reselect for this one to prepare the data before arriving to this component, you can achieve that with something like this :

```js
// VolumePage.js
const volumeListData = useSelector((state) => makeVolumeList(state, match));
```

```js
// src/services/utils.js
const getVolumeFromUrl = createSelector(
  getVolumeNameFromUrl,
  getVolumes,
  (volumeName, volumes) => {
    return volumes.find(
      (volume) =>
        volume && volume.metadata && volume.metadata.name === volumeName,
    );
  },
);

const makeGetNodeFromUrl2 = createSelector(
  (state, match) => {
    if (match && match.params && match.params.id) {
      return match.params.id;
    } else {
      return '';
    }
  },
  getNodes,
  (nodeName, nodes) => nodes.find((node) => node.name === nodeName) || {},
);

export const makeTestVolumes = createSelector(
  getVolumeFromUrl,
  getVolumes,
  (state) => state.app.volumes.pVList,
  (state) => state.app.volumes.pVCList,
  (state) => state.app.monitoring.volumeStats.volumeUsedCurrent,
  (state) => state.app.monitoring.alert,
  makeGetNodeFromUrl2,
  (state) => state.app.monitoring.volumeStats.volumeLatencyCurrent,
  (
    volume,
    volumes,
    pVList,
    pVCList,
    volumeUsedCurrentList,
    alerts,
    node,
    volumeLatencyCurrent,
  ) => {
    // console.log('makeTestVolumes');
    if (!volume) {
      return [];
    }

    return volumes.map((volume) => {
      const volumePV = pVList.find(
        (pV) => pV.metadata.name === volume.metadata.name,
      );
      // find the mapping PVC of this specific volume
      const volumePVC = pVCList.find(
        (pVC) => pVC.spec.volumeName === volume.metadata.name,
      );

      const volumeComputedCondition = computeVolumeCondition(
        computeVolumeGlobalStatus(volume.metadata.name, volume?.status),
        volumePV?.status?.phase === STATUS_BOUND
          ? intl.translate('yes')
          : intl.translate('no'),
      );

      let volumeUsedCurren = null;
      let volumeAlerts = [];
      let volumeHealth = '';

      // if volume is bounded
      if (volumePVC) {
        volumeUsedCurren = volumeUsedCurrentList?.find(
          (volStat) =>
            volStat.metric.persistentvolumeclaim === volumePVC.metadata.name,
        );

        // filter the alerts related to the current volume.
        volumeAlerts = alerts?.list?.filter(
          (alert) =>
            alert.labels.persistentvolumeclaim === volumePVC.metadata.name,
        );
      } else {
        volumeHealth = STATUS_NONE;
      }

      // THE RULES TO COMPUTE THE HEALTH
      // compute the volume health based on the severity of the alerts
      // critical => if there is at least one critical
      if (volumeAlerts.length) {
        volumeHealth = volumeAlerts.find(
          (vol) => vol.labels.severity === STATUS_CRITICAL,
        )
          ? STATUS_CRITICAL
          : STATUS_WARNING;
      } else if (volumeComputedCondition === ('exclamation' || 'unlink')) {
        volumeHealth = STATUS_NONE;
      } else {
        volumeHealth = STATUS_HEALTH;
      }

      return {
        name: volume.metadata.name,
        node: node.name,
        usage: volumeUsedCurren?.value[1]
          ? Math.round(
              (volumeUsedCurren?.value[1] /
                (volumePV?.spec?.capacity?.storage &&
                  allSizeUnitsToBytes(volumePV?.spec?.capacity?.storage))) *
                100,
            )
          : 0,
        status: volumeComputedCondition,
        bound:
          volumePV?.status?.phase === STATUS_BOUND
            ? intl.translate('yes')
            : intl.translate('no'),
        storageCapacity:
          volumePV?.spec?.capacity?.storage || intl.translate('unknown'),
        storageClass: volume.spec.storageClassName,
        creationTime: formatDate(volume.metadata.creationTimestamp),
        usageRawData: volumeUsedCurren?.value[1]
          ? bytesToSize(volumeUsedCurren?.value[1])
          : 0,
        health: volumeHealth,
        // when the data is not ready, display `loading` for the moment
        latency: volumeLatencyCurrent
          ? Math.round(
              volumeLatencyCurrent?.find(
                (vLV) => vLV.metric.device === volume.status.deviceName,
              )?.value[1] * 100,
            ) + 'ms'
          : 'loading',
      };
    });
  },
);
```

Please note that this is a quick draft, there might need some refinement. But that way it's already "easier" to test this critical part of the page (basically most login behind you `Table`.

Reply:
I extracted the `getVolumeListData` function to `NodeVolumesUtils.js` since it's not a global utility function yet.